oxia-operator/controllers/oxiacluster_coordinator.go

```diff
@@ -0,0 +1,218 @@
+// Copyright 2023 StreamNative, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controllers
+
+import (
+	"context"
+	monitoringV1 "github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring/v1"
+	oxiav1alpha1 "github.com/streamnative/oxia/api/v1alpha1"
+	appsV1 "k8s.io/api/apps/v1"
+	coreV1 "k8s.io/api/core/v1"
+	rbacV1 "k8s.io/api/rbac/v1"
+	apiErrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+)
+
+type ServerAddress struct {
+	// Public is the endpoint that is advertised to clients
+	Public string `json:"public" yaml:"public"`
+	// Internal is the endpoint for server->server RPCs
+	Internal string `json:"internal" yaml:"internal"`
+}
+
+type ClusterConfig struct {
+	Namespaces []NamespaceConfig `json:"namespaces" yaml:"namespaces"`
+	Servers    []ServerAddress   `json:"servers" yaml:"servers"`
+}
+
+type NamespaceConfig struct {
+	Name              string `json:"name" yaml:"name"`
+	InitialShardCount uint32 `json:"initialShardCount" yaml:"initialShardCount"`
+	ReplicationFactor uint32 `json:"replicationFactor" yaml:"replicationFactor"`
+}
+
+var CoordinatorPorts = []NamedPort{InternalPort, MetricsPort}
+
+func (r *OxiaClusterReconciler) reconcileCoordinator(ctx context.Context, oxia *oxiav1alpha1.OxiaCluster) (ctrl.Result, error) {
+	// Reconcile service account
+	_serviceAccount := &coreV1.ServiceAccount{}
+	nsName := types.NamespacedName{Name: resourceName(Coordinator, oxia.Name), Namespace: oxia.Namespace}
+	err := r.Client.Get(ctx, nsName, _serviceAccount)
+	if err != nil {
+		if !apiErrors.IsNotFound(err) {
+			r.Log.Error(err, "Failed to get service account.")
+			return ctrl.Result{}, err
+		}
+		r.Log.Info("Creating a new service account.", "NamespaceName", nsName)
+		err := r.Client.Create(ctx, serviceAccount(Coordinator, oxia))
+		if err != nil {
+			r.Log.Error(err, "Failed to create service account :", "NamespaceName", nsName)
+			return ctrl.Result{}, err
+		}
+		r.Log.Info("Created service account: ", "NamespaceName", nsName)
```

Comment:
Should consideration be given to logging the status of these resources to reflect the current progress of oxiaCluster object creation